presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/TestPushProjectionIntoTableScan.java

```diff
@@ -0,0 +1,247 @@
+    private static final Type ROW_TYPE = RowType.from(asList(field("a", BIGINT), field("b", BIGINT)));
+
+    private static final TableHandle TEST_TABLE_HANDLE = createTableHandleForMockConnector(TEST_SCHEMA, TEST_TABLE);
+
+    private static final Session MOCK_SESSION = testSessionBuilder().setCatalog(MOCK_CATALOG).setSchema(TEST_SCHEMA).build();
+
+    @Test
+    public void testDoesNotFire()
+    {
+        try (RuleTester ruleTester = new RuleTester()) {
+            String inputColumnName = "input_column";
+            Type inputColumnType = ROW_TYPE;
+            ColumnHandle inputColumnHandle = new TpchColumnHandle(inputColumnName, inputColumnType);
+
+            MockConnectorFactory factory = createMockFactory(ImmutableMap.of(inputColumnName, inputColumnHandle), false);
+
+            ruleTester.createCatalog(MOCK_CATALOG, factory, ImmutableMap.of());
+
+            ruleTester.assertThat(new PushProjectionIntoTableScan(ruleTester.getMetadata(), new TypeAnalyzer(new SqlParser(), ruleTester.getMetadata())))
+                .on(p -> {
+                    Symbol symbol = p.symbol(inputColumnName, inputColumnType);
+                    return p.project(
+                        Assignments.of(p.symbol("symbol_dereference", BIGINT), new DereferenceExpression(symbol.toSymbolReference(), new Identifier("a"))),
+                        p.tableScan(TEST_TABLE_HANDLE, ImmutableList.of(symbol), ImmutableMap.of(symbol, inputColumnHandle)));
+                })
+                .withSession(MOCK_SESSION)
+                .doesNotFire();
+        }
+    }
+
+    @Test
+    public void testPushProjection()
+    {
+        try (RuleTester ruleTester = new RuleTester()) {
+            // Building context for input
+            String inputColumnName = "col0";
+            Type inputColumnType = ROW_TYPE;
+            Symbol baseColumnSymbol = new Symbol(inputColumnName);
+            ColumnHandle columnHandle = new TpchColumnHandle(inputColumnName, inputColumnType);
+            ConnectorTableHandle expectedTableHandle = createTableHandleForMockConnector(TEST_SCHEMA, "projected_" + TEST_TABLE).getConnectorHandle();
+
+            // Create catalog with applyProjection enabled
+            MockConnectorFactory factory = createMockFactory(ImmutableMap.of(inputColumnName, columnHandle), true);
+            ruleTester.createCatalog(MOCK_CATALOG, factory, ImmutableMap.of());
+
+            Metadata metadata = ruleTester.getMetadata();
+            TypeAnalyzer typeAnalyzer = new TypeAnalyzer(new SqlParser(), metadata);
+            PushProjectionIntoTableScan optimizer = new PushProjectionIntoTableScan(metadata, typeAnalyzer);
+
+            // Prepare project node symbols and types
+            Symbol identitySymbol = new Symbol("symbol_identity");
+            Symbol dereferenceSymbol = new Symbol("symbol_dereference");
+            Symbol constantSymbol = new Symbol("symbol_constant");
```

Comment:
I guess you should use `PlanBuilder` to create symbols with types. Somewhere after:
```
            ruleTester.assertThat(optimizer)
                    .on(p -> {
```